### PR TITLE
Add regression tests for duplicate-index gather multiply (#3442)

### PR DESCRIPTION
**Pull request by Transurgeon**
## Description
Regression tests for #3442: a DNLP problem containing `cp.multiply(a[idx], b[idx])` where `idx` has **duplicate** entries (the issue's 9-bus power flow indexes bus variables by branch, `rows = [0, 0, 0, 1, ...]`) corrupted the heap during derivative-structure initialization and segfaulted, while the scalar-indexed loop formulation worked.

The root cause was in the diff engine, not cvxpy: `sparse_index_alloc` sized the gather Jacobian by the *source's* nnz, an undercount when duplicated indices select the same row more than once (`x[[0,0,1,2,2,3]]` on a size-4 variable gathers 6 entries from a 4-nnz source). Fixed in SparseDifferentiation/SparseDiffEngine#105, shipping in sparsediffpy 0.6.0.

This PR adds the cvxpy-side coverage:

- `TestDuplicateIndexGathers::test_multiply_duplicate_index_gathers_derivatives` — solver-free `DerivativeChecker` run over the issue's shape (`multiply(v[rows], v[cols])` and `nlp.cos(theta[rows] - theta[cols])`), cross-checking values, gradient, Jacobian, and Hessian against finite differences.
- `TestAffineDiffEngine::test_multiply_duplicate_index_gathers_solve` — IPOPT solve of the array-indexed formulation, asserted to agree with the scalar-indexed loop formulation (the issue's workaround).

Both are gated on `sparsediffpy >= 0.6.0`: on the currently pinned 0.5.x the bug corrupts the process heap, so the tests skip loudly (naming the engine fix) rather than crash CI. They activate automatically when the sparsediffpy pin is bumped.

Issue link: #3442

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [x] Bug fix
- [ ] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [x] Add our license to new files.
- [x] Check that your code adheres to our coding style.
- [x] Write unittests.
- [x] Run the unittests and check that they're passing.
- [ ] Run the benchmarks to make sure your change doesn't introduce a regression.

(Unittests run against a dev build of the fixed engine: the derivative-checker test passes locally. The IPOPT solve test has NOT been run locally — IPOPT is not installed here — it skips locally and will first be exercised by the NLP CI workflow once the sparsediffpy pin is bumped. No benchmark run — test-only change, no library code touched.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: cvxpy/tests/nlp_tests/stress_tests_diff_engine/test_affine_vector_atoms.py [hunk 1]
```diff
@@ -21,6 +21,51 @@
...
+    from importlib.metadata import version as _pkg_version
+    _ENGINE_FIXED = tuple(
+        int(p) for p in _pkg_version("sparsediffpy").split(".")[:2]) >= (0, 6)
+except Exception:  # pragma: no cover - not installed / unparsable version
+    _ENGINE_FIXED = False
+
+_ENGINE_SKIP_REASON = (
+    'requires sparsediffpy >= 0.6.0 (duplicate-index gather fix, '
+    'SparseDiffEngine#105).'
+)
```
Comment (Transurgeon) [inline cvxpy/tests/nlp_tests/stress_tests_diff_engine/test_affine_vector_atoms.py]: I'll clean this PR up once we make a new release of the diffengine.

**Review by PTNobel — APPROVED** [review]
My preference is to merge this PR after the SparseDiff release with the simplifications.

**Comment by Transurgeon** [comment]
> My preference is to merge this PR after the SparseDiff release with the simplifications.

yes of course, that's what I had in mind too. We should ensure the CI passes as well.

File: cvxpy/tests/nlp_tests/stress_tests_diff_engine/test_affine_vector_atoms.py (hunk 1 above)
Comment (Transurgeon) [reply cvxpy/tests/nlp_tests/stress_tests_diff_engine/test_affine_vector_atoms.py]: Done: sparsediffpy 0.6.0 (with the SparseDiffEngine#105 gather fix) is on PyPI, so the `_ENGINE_FIXED` version sniffing and both skip markers are gone — the pin floor is now `>= 0.6.0, < 0.7.0` and the tests run unconditionally.

**Comment by Transurgeon** [comment]
sparsediffpy **0.6.0 is released on PyPI**, so this is cleaned up as promised: the version-sniffing gate is removed, the requirement is bumped to `sparsediffpy >= 0.6.0, < 0.7.0` (same one-liner as #3447; whichever merges second sees a no-op), and CI now actually exercises the regression tests. The solver-free DerivativeChecker test passes locally against the released wheel; the IPOPT solve test runs in the `run_nlp_tests` job here.
